src/serialize/text/index.ts

```diff
@@ -1,10 +1,91 @@
-import { DOMSerializer, Node as ProsemirrorNode } from 'prosemirror-model';
-import { getSchema, UseSchema } from '../../schemas';
-
-export function toText(doc: ProsemirrorNode, useSchema: UseSchema, document: Document) {
-  const div = document.createElement('div');
-  const schema = getSchema(useSchema);
-  const frag = DOMSerializer.fromSchema(schema).serializeFragment(doc.content, { document });
-  div.appendChild(frag);
-  return div.textContent ?? '';
+import { MarkdownSerializer, MarkdownSerializerState } from 'prosemirror-markdown';
+import { Node as ProsemirrorNode } from 'prosemirror-model';
+import * as nodes from '../../nodes';
+
+function simpleInlineRender(state: MarkdownSerializerState, node: ProsemirrorNode) {
+  state.renderInline(node);
+}
+
+function simpleBlockRender(state: MarkdownSerializerState, node: ProsemirrorNode) {
+  state.renderInline(node);
+  state.closeBlock(node);
+}
+
+function newLine(state: MarkdownSerializerState) {
+  state.ensureNewLine();
+}
+
+function nothing() {
+  // pass
+}
+
+const noMarks = {
+  open: '',
+  close: '',
+  mixable: true,
+};
+
+export const textSerializer = new MarkdownSerializer(
+  {
+    text(state, node) {
+      state.text(node.text ?? '', false);
+    },
+    paragraph: simpleBlockRender,
+    heading: simpleBlockRender,
+    blockquote: simpleBlockRender,
+    code_block: simpleBlockRender,
+    horizontal_rule: newLine,
+    hard_break: newLine,
+    ordered_list(state, node) {
+      const start = node.attrs.order || 1;
+      const maxW = String(start + node.childCount - 1).length;
+      const space = state.repeat(' ', maxW + 2);
+      state.renderList(node, space, (i) => {
+        const nStr = String(start + i);
+        return `${state.repeat(' ', maxW - nStr.length) + nStr}. `;
+      });
+    },
+    bullet_list(state, node) {
+      state.renderList(node, '  ', () => `${node.attrs.bullet || '*'} `);
+    },
+    list_item(state, node) {
+      state.renderInline(node);
+    },
+    // Presentational
+    image: newLine,
+    iframe: newLine,
+    time: nodes.Time.toMarkdown, // This is just the time! :)
+    callout: simpleBlockRender,
+    aside: simpleBlockRender,
+    // Technical
+    math: simpleInlineRender,
+    equation: simpleBlockRender,
+    cite: nodes.Cite.toMarkdown, // TODO: fix this
+    cite_group: nodes.CiteGroup.toMarkdown, // TODO: fix this
+    // Tables
+    table: nodes.Table.toMarkdown,
+    // Dynamic
+    variable: newLine,
+    display: nodes.Display.toMarkdown, // TODO: fix this
+    dynamic: nodes.Dynamic.toMarkdown, // TODO: fix this
```

Comment:
There are some improvements that we can make here, but will do that in another pass.